File: src/rulers.ts
/**
 * Rulers: how performance is measured.
 *
 * A grader judges one trial. A ruler turns the rows of many trials into one
 * estimate with a 95% interval: accuracy, a cost, pass^k, agreement between
 * raters. Rulers are the vocabulary of an experiment; certificates use them too.
 */

import { ValidationError } from "./models.js";
import { type Estimate, type Row, type Rows, estimateMetric, interval, scenarioGroups } from "./statistics/core.js";
import { mean, quantile } from "./statistics/dist.js";
import { Rng } from "./statistics/random.js";

export type RulerResult = Partial<Estimate> & { estimate: number | null; interval: [number | null, number | null]; ruler?: string; [extra: string]: unknown };
export type Measure = (rows: Rows, bootstrapSamples: number, seed: number) => RulerResult;

export class Ruler {
  readonly name: string;
  readonly measure: Measure;
  readonly description: string;
  readonly metric: string | null;
  readonly higherIsBetter: boolean | null;

  constructor(name: string, measure: Measure, options: { description?: string; metric?: string | null; higherIsBetter?: boolean | null } = {}) {
    this.name = name;
    this.measure = measure;
    this.description = options.description ?? "";
    this.metric = options.metric ?? null;
    this.higherIsBetter = options.higherIsBetter ?? null;
    Object.freeze(this);
  }

  apply(rows: Rows, options: { bootstrapSamples?: number; seed?: number } = {}): RulerResult {
    const result = { ...this.measure(rows, options.bootstrapSamples ?? 2000, options.seed ?? 0) };
    result.ruler ??= this.name;
    return result;
  }
}

const LOWER_IS_BETTER = new Set(["cost", "latencyMs", "tokens", "steps", "retries"]);

/**
 * Scenario-weighted mean of a row metric with a top-cluster bootstrap interval.
 * Metrics: correct, fieldF1, schemaValid, cost, latencyMs, tokens, steps, retries,
 * attempts:<harm>, occurrences:<harm>, loss:<harm>, metric:<name>.
 */
export function rate(metric: string, options: { name?: string; higherIsBetter?: boolean | null; description?: string } = {}): Ruler {
  let higher = options.higherIsBetter;
  if (higher === undefined) {
    higher = ["correct", "fieldF1", "schemaValid"].includes(metric) ? true : null;
    if (LOWER_IS_BETTER.has(metric) || metric.startsWith("attempts:") || metric.startsWith("occurrences:") || metric.startsWith("loss:")) higher = false;
  }
  return new Ruler(options.name ?? metric, (rows, samples, seed) => estimateMetric(rows, metric, { samples, seed }),
    { description: options.description ?? `Mean ${metric} per scenario, 95% cluster interval`, metric, higherIsBetter: higher });
}

export function accuracy(): Ruler {
  return rate("correct", { name: "accuracy", description: "Fraction of scenarios with the correct final state" });
}

function perScenario(rows: Rows, statistic: (trials: Row[]) => number, samples: number, seed: number): Estimate {
  const groups = [...scenarioGroups(rows).values()];
  return interval(groups.map(statistic), groups.map((trials) => trials[0].cluster), { samples, seed, bounds: [0, 1] });
}

export function passPowerK(): Ruler {
  return new Ruler("pass^k", (rows, s, seed) => perScenario(rows, (trials) => (trials.every((r) => r.correct) ? 1 : 0), s, seed),
    { description: "Fraction of scenarios where every repetition is correct", higherIsBetter: true });
}

export function pathConsistency(): Ruler {
  return new Ruler("path_consistency", (rows, s, seed) => perScenario(rows, (trials) => (new Set(trials.map((r) => r.pathSignature)).size === 1 ? 1 : 0), s, seed),
    { description: "Fraction of scenarios whose repetitions took the identical tool-call path", higherIsBetter: true });
}

/** Nominal Krippendorff's alpha over units, each a list of ratings from different raters. */
export function krippendorffAlpha(units: readonly (readonly string[])[]): number | null {
  const pairable = units.filter((u) => u.length >= 2);
  const n = pairable.reduce((sum, u) => sum + u.length, 0);
  if (n < 2) return null;
  const totals = new Map<string, number>();
  for (const unit of pairable) for (const v of unit) totals.set(v, (totals.get(v) ?? 0) + 1);
  let observed = 0;
  for (const unit of pairable) {
    const counts = new Map<string, number>();
    for (const v of unit) counts.set(v, (counts.get(v) ?? 0) + 1);
    let disagreements = 0;
    for (const [c, nc] of counts) for (const [k, nk] of counts) if (c !== k) disagreements += nc * nk;
    observed += disagreements / (unit.length - 1);
  }
  observed /= n;
  let expected = 0;
  for (const [c, nc] of totals) for (const [k, nk] of totals) if (c !== k) expected += nc * nk;
  expected /= n * (n - 1);
  if (expected === 0) return observed === 0 ? 1 : null;
  return 1 - observed / expected;
}

function clusterBootstrap(rows: Rows, statistic: (rows: Rows) => number | null, samples: number, seed: number, bounds: [number, number] | null): RulerResult {
  const estimate = statistic(rows);
  const byCluster = new Map<string, Row[]>();
  for (const row of rows) { const list = byCluster.get(row.cluster) ?? []; list.push(row); byCluster.set(row.cluster, list); }
  const names = [...byCluster.keys()].sort();
  const result: RulerResult = { estimate, interval: [null, null], confidence: 0.95, clusters: names.length, scenarios: scenarioGroups(rows).size,
    method: "top_cluster_bootstrap_percentile" };
  if (estimate === null || names.length < 2) { result.method = "not_estimable; need the statistic and at least two clusters"; return result; }
  const rng = new Rng(seed);
  const draws: number[] = [];
  for (let s = 0; s < samples; s++) {
    const chosen: Row[] = [];
    for (let k = 0; k < names.length; k++) chosen.push(...byCluster.get(names[rng.integer(0, names.length)])!);
    const value = statistic(chosen);
    if (value !== null) draws.push(value);
  }
  if (draws.length < samples / 2) { result.method = "not_estimable; statistic undefined in most resamples"; return result; }
  let low = quantile(draws, 0.025), high = quantile(draws, 0.975);
  if (bounds) { low = Math.max(bounds[0], low); high = Math.min(bounds[1], high); }
  result.interval = [low, high];
  return result;
}

/**
 * Krippendorff's alpha between named raters found in each trial's ratings.
 * Raters may be a deterministic grader, humans, or a language-model judge run
 * over the transcripts afterwards. Low agreement means the judge is not yet a ruler.
 */
export function interRaterReliability(raters: readonly string[]): Ruler {
  if (raters.length < 2) throw new ValidationError("Inter-rater reliability needs at least two raters");
  const statistic = (rows: Rows) => krippendorffAlpha(rows.map((r) => raters.filter((name) => name in (r.ratings ?? {})).map((name) => String(r.ratings[name]))));
  return new Ruler(`irr:${raters.join("+")}`, (rows, s, seed) => clusterBootstrap(rows, statistic, s, seed, [-1, 1]),
    { description: `Nominal Krippendorff's alpha between raters ${raters.join(", ")}`, higherIsBetter: true });
}

/** Fraction of trials where a rater's verdict (truthy = correct) matches the deterministic grader. */
export function agreementWithGrader(rater: string): Ruler {
  const statistic = (trials: Row[]) => {
    const judged = trials.filter((r) => rater in (r.ratings ?? {})).map((r) => (Boolean(r.ratings[rater]) === r.correct ? 1 : 0));
    return judged.length ? mean(judged) : NaN;
  };
  return new Ruler(`agreement:${rater}`, (rows, s, seed) => {
    const rated = rows.filter((r) => rater in (r.ratings ?? {}));
    if (!rated.length) return { estimate: null, interval: [null, null], method: "not_estimable; no ratings from this rater" };
    return perScenario(rated, statistic, s, seed);
  }, { description: `Agreement between rater ${rater} and the deterministic outcome grader`, higherIsBetter: true });
}

export const DEFAULT_RULERS: readonly Ruler[] = Object.freeze([accuracy(), rate("fieldF1"), passPowerK(), pathConsistency(), rate("cost"), rate("latencyMs"), rate("steps")]);

export function applyRulers(rows: Rows, rulers: readonly Ruler[], options: { functionId: string; bootstrapSamples?: number; seed?: number }): Record<string, RulerResult> {
  const selected = rows.filter((r) => r.functionId === options.functionId);
  if (!selected.length) throw new ValidationError(`No observations for function ${options.functionId}`);
  return Object.fromEntries(rulers.map((ruler) => [ruler.name, ruler.apply(selected, options)]));
}

File: src/signal_sdk/rulers.py
"""Rulers: how performance is measured.

A grader judges one trial. A ruler turns the rows of many trials into one
estimate with a 95% interval: accuracy, a cost, pass^k, agreement between
raters. Rulers are the vocabulary of an experiment; certificates use them too.
"""

from __future__ import annotations

from collections import Counter, defaultdict
from dataclasses import dataclass
from typing import Any, Callable, Mapping, Sequence

import numpy as np

from .statistics.core import Rows, trajectory_groups, estimate_metric, interval

Statistic = Callable[[Rows], float | None]


@dataclass(frozen=True)
class Ruler:
    name: str
    measure: Callable[[Rows, int, int], dict[str, Any]]
    description: str = ""
    metric: str | None = None
    higher_is_better: bool | None = None

    def __call__(self, rows: Rows, *, bootstrap_samples: int = 2000, seed: int = 0) -> dict[str, Any]:
        result = dict(self.measure(rows, bootstrap_samples, seed))
        result.setdefault("ruler", self.name)
        return result


def rate(metric: str, *, name: str | None = None, higher_is_better: bool | None = None,
         description: str = "") -> Ruler:
    """Trajectory-weighted mean of a row metric with a top-cluster bootstrap interval.

    Metrics: correct, field_f1, schema_valid, cost, latency_ms, tokens, steps, retries,
    attempts:<harm>, occurrences:<harm>, loss:<harm>, metric:<name>.
    """
    if higher_is_better is None:
        higher_is_better = metric in {"correct", "field_f1", "schema_valid"} or None
        if metric in {"cost", "latency_ms", "tokens", "steps", "retries"} or metric.startswith(("attempts:", "occurrences:", "loss:")):
            higher_is_better = False
    return Ruler(name or metric, lambda rows, samples, seed: estimate_metric(rows, metric, samples=samples, seed=seed),
                 description or f"Mean {metric} per trajectory, 95% cluster interval", metric, higher_is_better)


def accuracy() -> Ruler:
    return rate("correct", name="accuracy", description="Fraction of trajectories with the correct final state")


def _per_trajectory(rows: Rows, statistic: Callable[[list[Mapping[str, Any]]], float], samples: int, seed: int) -> dict[str, Any]:
    groups = trajectory_groups(rows)
    values = [statistic(trials) for trials in groups.values()]
    clusters = [str(trials[0]["cluster"]) for trials in groups.values()]
    return interval(values, clusters, samples=samples, seed=seed, bounds=(0.0, 1.0))


def pass_power_k() -> Ruler:
    return Ruler("pass^k", lambda rows, s, seed: _per_trajectory(rows, lambda t: float(all(r["correct"] for r in t)), s, seed),
                 "Fraction of trajectories where every repetition is correct", None, True)


def path_consistency() -> Ruler:
    return Ruler("path_consistency",
                 lambda rows, s, seed: _per_trajectory(rows, lambda t: float(len({r.get("path_signature") for r in t}) == 1), s, seed),
                 "Fraction of trajectories whose repetitions took the identical tool-call path", None, True)


def krippendorff_alpha(units: Sequence[Sequence[Any]]) -> float | None:
    """Nominal Krippendorff's alpha over units, each a list of ratings from different raters."""
    pairable = [tuple(u) for u in units if len(u) >= 2]
    n = sum(len(u) for u in pairable)
    if n < 2:
        return None
    totals = Counter(v for u in pairable for v in u)
    observed = 0.0
    for unit in pairable:
        counts = Counter(unit)
        observed += sum(counts[c] * counts[k] for c in counts for k in counts if c != k) / (len(unit) - 1)
    observed /= n
    expected = sum(totals[c] * totals[k] for c in totals for k in totals if c != k) / (n * (n - 1))
    if expected == 0:
        return 1.0 if observed == 0 else None
    return 1.0 - observed / expected


def _cluster_bootstrap(rows: Rows, statistic: Statistic, samples: int, seed: int,
                       bounds: tuple[float, float] | None = None) -> dict[str, Any]:
    estimate = statistic(rows)
    by_cluster: dict[str, list[Mapping[str, Any]]] = defaultdict(list)
    for row in rows:
        by_cluster[str(row["cluster"])].append(row)
    names = sorted(by_cluster)
    result = {"estimate": estimate, "interval": [None, None], "confidence": 0.95,
              "clusters": len(names), "trajectories": len(trajectory_groups(rows)),
              "method": "top_cluster_bootstrap_percentile"}
    if estimate is None or len(names) < 2:
        result["method"] = "not_estimable; need the statistic and at least two clusters"
        return result
    rng = np.random.default_rng(seed)
    draws = []
    for _ in range(samples):
        chosen = rng.integers(0, len(names), len(names))
        value = statistic([r for i in chosen for r in by_cluster[names[i]]])
        if value is not None:
            draws.append(value)
    if len(draws) < samples // 2:
        result["method"] = "not_estimable; statistic undefined in most resamples"
        return result
    low, high = np.quantile(draws, [0.025, 0.975])
    if bounds:
        low, high = max(bounds[0], low), min(bounds[1], high)
    result["interval"] = [float(low), float(high)]
    return result


def inter_rater_reliability(raters: Sequence[str]) -> Ruler:
    """Krippendorff's alpha between named raters found in each trial's ratings.

    Raters may be a deterministic grader, humans, or a language-model judge run over
    the transcripts afterwards. Low agreement means the judge is not yet a ruler.
    """
    raters = tuple(raters)
    if len(raters) < 2:
        raise ValueError("Inter-rater reliability needs at least two raters")

    def statistic(rows: Rows) -> float | None:
        units = [[str(r["ratings"][name]) for name in raters if name in r.get("ratings", {})] for r in rows]
        return krippendorff_alpha(units)

    return Ruler("irr:" + "+".join(raters),
                 lambda rows, s, seed: _cluster_bootstrap(rows, statistic, s, seed, bounds=(-1.0, 1.0)),
                 f"Nominal Krippendorff's alpha between raters {', '.join(raters)}", None, True)


def agreement_with_grader(rater: str) -> Ruler:
    """Fraction of trials where a rater's verdict (truthy = correct) matches the deterministic grader."""
    def statistic(trials: list[Mapping[str, Any]]) -> float:
        judged = [(bool(r["ratings"][rater]) == bool(r["correct"])) for r in trials if rater in r.get("ratings", {})]
        return float(np.mean(judged)) if judged else float("nan")

    def measure(rows: Rows, samples: int, seed: int) -> dict[str, Any]:
        rated = [r for r in rows if rater in r.get("ratings", {})]
        if not rated:
            return {"estimate": None, "interval": [None, None], "method": "not_estimable; no ratings from this rater"}
        return _per_trajectory(rated, statistic, samples, seed)

    return Ruler(f"agreement:{rater}", measure, f"Agreement between rater {rater} and the deterministic outcome grader", None, True)


DEFAULT_RULERS: tuple[Ruler, ...] = (accuracy(), rate("field_f1"), pass_power_k(), path_consistency(),
                                    rate("cost"), rate("latency_ms"), rate("steps"))


def apply_rulers(rows: Rows, rulers: Sequence[Ruler], *, function_id: str,
                 bootstrap_samples: int = 2000, seed: int = 0) -> dict[str, dict[str, Any]]:
    selected = [r for r in rows if str(r["function_id"]) == function_id]
    if not selected:
        raise ValueError(f"No observations for function {function_id}")
    return {ruler.name: ruler(selected, bootstrap_samples=bootstrap_samples, seed=seed) for ruler in rulers}
